src/NuGet.Core/NuGet.DependencyResolver.Core/GraphModel/Tracker.cs

```diff
@@ -3,83 +3,139 @@
         }
 
-        private Entry GetEntry(GraphItem<TItem> item)
+        private Entry? TryGetEntry(GraphItem<TItem> item)
         {
-            Entry itemList;
-            if (!_entries.TryGetValue(item.Key.Name, out itemList))
+            _entryByLibraryName.TryGetValue(item.Key.Name, out Entry? entry);
+            return entry;
+        }
+
+        private Entry GetOrAddEntry(GraphItem<TItem> item)
+        {
+            if (!_entryByLibraryName.TryGetValue(item.Key.Name, out Entry? entry))
             {
-                itemList = new Entry();
-                _entries[item.Key.Name] = itemList;
+                entry = new Entry();
+                _entryByLibraryName[item.Key.Name] = entry;
             }
-            return itemList;
+
+            return entry;
         }
 
-        private class Entry
+        private sealed class Entry
         {
-            public Entry()
+            /// <summary>
+            /// This field can have one of three values:
+            ///
+            /// - null: when no graph items exist in this entry
+            /// - a graph item: when only a single graph item exists in this entry
+            /// - a hash set of graph items: when multiple graph items exist in this entry
+            ///
+            /// This packing exists in order to reduce the size of this object in
+            /// memory. For large graphs, this can amount to a significant saving in memory,
+            /// which helps overall performance.
+            /// </summary>
+            private object? _storage;
+
+            public bool Ambiguous { get; set; }
+
+            public void AddItem(GraphItem<TItem> item)
             {
-                List = new HashSet<GraphItem<TItem>>();
+                if (_storage is null)
+                {
+                    _storage = item;
+                }
+                else if (_storage is GraphItem<TItem> existingItem)
+                {
+                    if (!existingItem.Equals(item))
+                    {
+#if NETSTANDARD2_0
+                        _storage = new HashSet<GraphItem<TItem>>() { existingItem, item };
+#else
+                        _storage = new HashSet<GraphItem<TItem>>(capacity: 3) { existingItem, item };
+#endif
+                    }
+                }
+                else
+                {
+                    ((HashSet<GraphItem<TItem>>)_storage).Add(item);
+                }
             }
 
-            public HashSet<GraphItem<TItem>> List { get; set; }
+            public bool HasMultipleItems => _storage is HashSet<GraphItem<TItem>>;
 
-            public bool Ambiguous { get; set; }
+            public IEnumerable<GraphItem<TItem>> Items
+            {
+                get
+                {
+                    if (_storage is null)
+                        return Enumerable.Empty<GraphItem<TItem>>();
+                    if (_storage is GraphItem<TItem> item)
+                        return new[] { item };
```

Comment:
This returns an array of one element.

Reply:
Thank you. [docs](https://learn.microsoft.com/dotnet/csharp/programming-guide/arrays/single-dimensional-arrays#array-initialization) clearly states that `The length specifier isn't needed because it's inferred by the number of elements in the initialization list.`

```cs
string[] weekDays = new string[] { "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat" };
```

Reply:
Arrays can't differentiate between count and capacity. Collections like List and hash based containers do, where they allocate oversized arrays internally.